### PR TITLE
7377 password reset and login instructions

### PR DESCRIPTION
**Pull request by scttnlsn**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -405,16 +405,14 @@ def must_have_password_reset_on_create
       end
     end
 
-    def password_reset_cant_be_email_if_no_email
-      if reset_password_method == "email" && email.blank?
-        verb = new_record? ? "send" : "reset"
-        errors.add(:reset_password_method, :cant_passwd_email, verb: verb)
-      end
+    def must_have_password_on_enter
```
[the remark this answers is not in the document]
Comment (scttnlsn) [reply app/models/user.rb]: Not sure what the issue is here

**Comment by scttnlsn** [comment]
@smoyth Regarding the password disappearing on password validation...this is the behavior on staging as well (and for Rails' `password_field`s).  If you want this behavior to change we could change this line: https://github.com/thecartercenter/elmo/blob/develop/app/helpers/elmo_form_builder.rb#L182

```ruby
password_field(field_name, class: "text form-control", value: options[:value])
```

Similarly in the view we'd need to explicitly pass the value:

```ruby
f.field(:password, type: :password, value: f.object.password)
```

Password field docs: https://apidock.com/rails/ActionView/Helpers/FormHelper/password_field

**Comment by smoyth** [comment]
Oh ok, whoops! Let's leave the password field behavior alone then.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -60,16 +60,15 @@ def update
     pref_lang_changed = @user.pref_lang_changed?
 
     if @user.save
+      # if the user's password was reset, do it, and show instructions if requested
+      @user.reset_password_if_requested
```
[the remark this answers is not in the document]
Comment (scttnlsn) [reply app/controllers/users_controller.rb]: I still need to make this refactor but I agree- sending emails from the model is a bit of an anti-pattern.  I'm not testing that a specific email is being sent but rather just that it triggers some kind of email delivery:

```
expect { click_button("Save") }.to change { ActionMailer::Base.deliveries.count }.by 1
```

Does this seem sufficient?  I could also probably check that the delivered mail has the expected subject or contents.

**Comment by scttnlsn** [comment]
@smoyth This is ready for another review.  I *think* I addressed everything you mentioned, though there was a lot so please let me know if something needs more attention.

**Pull request merged**